File: src/main/java/net/hearthstats/HearthstoneMatch.java
package net.hearthstats;

import net.hearthstats.util.Rank;
import org.apache.commons.lang3.StringUtils;
import org.json.simple.JSONObject;

public class HearthstoneMatch {

	private String _mode;
	private String _userClass;
	private String _opponentClass;
	private boolean _coin = false;
	private String _result;
	private int _deckSlot;
	private String _opponentName;
	private Rank rankLevel;
	private int _numTurns = 0;
	private int _duration;
	private String _notes;
	private Integer _id;
	
	public HearthstoneMatch() {
		
	}
	public int getDeckSlot() {
		return _deckSlot;
	}
	
	public int getDuration() {
		return _duration;
	}
	
	public String getMode() {
		return _mode;
	}
	
	public int getNumTurns() {
		return _numTurns;
	}

	public void setMode(String mode) {
		_mode = mode;
	}
	
	public void setNumTurns(int numTurns) {
		_numTurns = numTurns;
	}

	public String getUserClass() {
		return _userClass;
	}

	public void setUserClass(String userClass) {
		_userClass = userClass;
	}

	public String getOpponentClass() {
		return _opponentClass;
	}

	public void setOpponentClass(String opponentClass) {
		_opponentClass = opponentClass;
	}
	
	public void setOpponentName(String opponentName) {
		_opponentName = opponentName;
	}
	
	public String getOpponentName() {
		return _opponentName;
	}

	public boolean hasCoin() {
		return _coin;
	}

	public void setCoin(boolean coin) {
		_coin = coin;
	}
	
	public void setDeckSlot(int deckSlot) {
		_deckSlot = deckSlot;
	}
	
	public void setDuration(int duration) {
		_duration = duration;
	}

	public Rank getRankLevel() {
		return rankLevel;
	}
	public String getResult() {
		return _result;
	}

	public void setRankLevel(Rank rankLevel) {
		this.rankLevel = rankLevel;
	}
	
	public void setResult(String result) {
		_result = result;
	}
	
	private String _propertyOrUnknown(String propertyVal) {
		return propertyVal == null ? "[undetected]" : propertyVal;
	}
	public String toString() {
		JSONObject deck = DeckUtils.getDeckFromSlot(getDeckSlot());
		
		return _propertyOrUnknown(getMode()) + 
				("Ranked".equals(getMode()) ? " level " + getRankLevel() : "") +
				" " +
				(hasCoin() ? "" : "no ") + "coin " + 
				_propertyOrUnknown(getUserClass()) + " vs. " +
				_propertyOrUnknown(getOpponentClass()) + " " +
				"(" + _propertyOrUnknown(getOpponentName()) + ") " +
				getResult() + " " +
				(deck == null ? "" : " deck: " + deck.get("name")) +
				" (" + getNumTurns() + " turns)";
				
	}

	@SuppressWarnings("unchecked")
	public JSONObject toJsonObject() {
		String result = null;
		switch(getResult()) {
			case "Victory": case "Win":
				result = "Win";
				break;
			case "Defeat": case "Loss":
				result = "Loss";
				break;
			case "Draw":
				result = "Draw";
				break;
		}
		JSONObject obj = new JSONObject();

		obj.put("mode", getMode());
        if ("Ranked".equals(getMode())) {
            if (getRankLevel() == Rank.LEGEND) {
                obj.put("ranklvl", 26);     // 26 currently represents legend on HearthStats
                obj.put("legend", "true");
            } else {
                obj.put("ranklvl", getRankLevel().number);
                obj.put("legend", "false");
            }
        }

        obj.put("slot", getDeckSlot());
		obj.put("class", getUserClass());
		obj.put("oppclass", getOpponentClass());
		obj.put("oppname", getOpponentName());
		obj.put("coin", hasCoin() ? "true" : "false");
        obj.put("result", result);
		obj.put("notes", getNotes());
		obj.put("numturns", getNumTurns());
		obj.put("duration", getDuration());
		
		return obj;
	}

    /**
     * Determines if the data for this match is complete.
     *
     * @return true if there is enough data to submit the match, false if some data is missing
     */
    public boolean isDataComplete() {
        if (getMode() == null) {
            return false;
        } else if (getResult() == null) {
            return false;
        } else if (getUserClass() == null) {
            return false;
        } else if (getOpponentClass() == null) {
            return false;
        } else if (StringUtils.isBlank(getOpponentName())) {
            return false;
        }

        if (getMode().equals("Ranked")) {
            if (getRankLevel() == null) {
                return false;
            } else if (getDeckSlot() < 1 || getDeckSlot() > 9) {
                return false;
            }
        } else if (getMode().equals("Casual")) {
            if (getDeckSlot() < 1 || getDeckSlot() > 9) {
                return false;
            }
        }

        return true;
    }

	public String getNotes() {
		return _notes;
	}
	public void setNotes(String text) {
		_notes = text;
	}
	
	public String getEditUrl() {
		return "http://hearthstats.net/constructeds/" + getId() + "/edit";		
	}

	public Integer getId() {
		return _id;
	}
	public void setId(int id) {
		_id  = id;
	}

}

File: uploader-clean/src/main/scala/net/hearthstats/core/HearthstoneMatch.scala
package net.hearthstats.core

import scala.collection.JavaConversions.mutableMapAsJavaMap
import org.apache.commons.lang3.StringUtils
import org.json.simple.JSONObject
import com.github.nscala_time.time.Imports.DateTime
import grizzled.slf4j.Logging

//TODO use options
//TODO avoid mutable 
class HearthstoneMatch(var mode: GameMode = GameMode.UNDETECTED,
  var userClass: HeroClass = HeroClass.UNDETECTED,
  var opponentClass: HeroClass = HeroClass.UNDETECTED,
  var coin: Option[Boolean] = None,
  var result: Option[MatchOutcome] = None,
  var deck: Option[Deck] = None,
  var opponentName: String = null,
  var rankLevel: Option[Rank] = None,
  var numTurns: Int = -1,
  var duration: Int = -1,
  var notes: String = null,
  var id: Int = -1) extends Logging {

  debug("new HearthstoneMatch")

  private var _userClassUnconfirmed: Boolean = true
  //needed for java calls
  def this() = this(mode = null)
  
  val startedAt: Long = System.currentTimeMillis

  def describeResult: String = result match {
    case Some(r) => r.toString
    case None => "UnknownResult"
  }

  def deckSlot: Option[Int] = for {
    d <- deck
    s <- d.activeSlot
  } yield s

  def toJsonObject: JSONObject = {
    val map = collection.mutable.Map(
      "mode" -> mode.toString,
      "slot" -> deckSlot.getOrElse(-1),
      "class" -> userClass.toString,
      "oppclass" -> opponentClass.toString,
      "oppname" -> opponentName,
      "coin" -> coin.toString,
      "result" -> describeResult,
      "notes" -> notes,
      "numturns" -> numTurns,
      "duration" -> duration)

    if (mode == GameMode.RANKED) {
      if (Rank.LEGEND == rankLevel) {
        map += "ranklvl" -> 26
        map += "legend" -> "true"
      } else {
        map += "ranklvl" -> rankLevel.get.number
        map += "legend" -> "false"
      }
    }
    new JSONObject(map)
  }

  /**
   * Determines if the data for this match is complete.
   *
   * @return true if there is enough data to submit the match, false if some data is missing
   */
  def isDataComplete: Boolean =
    mandatoryFieldsOK &&
      (mode match {
        case GameMode.RANKED => rankLevel.isDefined && deckSlotOk
        case GameMode.CASUAL => deckSlotOk
        case _ => true
      })

  def deckSlotOk = deckSlot.isDefined &&
    deckSlot.get >= 1 && deckSlot.get <= 9

  def mandatoryFieldsOK =
    result.isDefined &&
      userClass != HeroClass.UNDETECTED &&
      opponentClass != HeroClass.UNDETECTED &&
      StringUtils.isNotBlank(opponentName) &&
      mode != GameMode.UNDETECTED

  def editUrl: String =
    s"http://hearthstats.net/constructeds/$id/edit"
    
  def endMatch = duration =  Math.round((System.currentTimeMillis - startedAt) / 1000)
}
